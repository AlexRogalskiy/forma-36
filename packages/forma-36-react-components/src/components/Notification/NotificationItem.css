@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';
@import 'settings/dimensions';

.NotificationItem {
  display: flex;
  padding: var(--spacing-s);
  padding-left: var(--spacing-m);
  transition: transform var(--transition-duration-default)
      var(--transition-easing-default),
    opacity var(--transition-duration-default) var(--transition-easing-default);
  transform: translateY(0);
  color: var(--color-white);
  cursor: default;
  font-family: var(--font-stack-primary);
  font-size: var(--font-size-m);
  font-weight: var(--font-weight-medium);
  line-height: var(--line-height-default);
}

.NotificationItem--success {
  background: var(--color-green-dark);
}

.NotificationItem--error {
  background: var(--color-red-dark);
}

.NotificationItem__intent {
  composes: sr-only from './../../styles/settings/helpers.css';
}

.NotificationItem__icon {
  margin-right: var(--spacing-s);
  display: flex;
  margin-top: 1px;
}

.NotificationItem__text {
  flex-grow: 1;
  margin-right: var(--spacing-s);
}

.NotificationItem__dismiss {
  background: transparent;
  border: none;
  cursor: pointer;
  outline: none;
  pointer-events: all;
}
