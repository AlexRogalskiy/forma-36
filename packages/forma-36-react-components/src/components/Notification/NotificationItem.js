import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import styles from './NotificationItem.css';
import IconButton from '../IconButton';
import Icon from '../Icon';
import { iconName } from '../Icon/constants';

const Icons = {
  success: iconName.CheckCircle,
  error: iconName.Warning,
};

export const NotificationItemPropTypes = {
  intent: PropTypes.oneOf(['error', 'success']),
  onClose: PropTypes.func.isRequired,
  hasCloseButton: PropTypes.bool,
  children: PropTypes.node.isRequired,
  testId: PropTypes.string,
};

class NotificationItem extends React.Component {
  static propTypes = NotificationItemPropTypes;

  static defaultProps = {
    testId: 'cf-ui-notification',
    intent: 'success',
    hasCloseButton: true,
  };

  render() {
    const { children, testId, intent, onClose, hasCloseButton } = this.props;

    const icon = Icons[intent];
    if (!icon) {
      throw new Error(`Intent ${intent} is not supported in Note component.`);
    }
    const classes = classNames(styles.NotificationItem, {
      [styles[`NotificationItem--${intent}`]]: true,
    });
    return (
      <div
        className={classes}
        data-test-id={testId}
        data-intent={intent}
        role="alert"
        aria-live={intent === 'success' ? 'polite' : 'assertive'}
      >
        <div className={styles.NotificationItem__intent}>{intent}</div>
        <div className={styles.NotificationItem__icon} aria-hidden="true">
          <Icon icon={icon} color="white" />
        </div>
        <div className={styles.NotificationItem__text}>{children}</div>
        {hasCloseButton && (
          <IconButton
            iconProps={{ icon: 'Close' }}
            onClick={onClose}
            testId="cf-ui-notification-close"
            label="Dismiss"
            className={styles.NotificationItem__dismiss}
          />
        )}
      </div>
    );
  }
}

export default NotificationItem;
