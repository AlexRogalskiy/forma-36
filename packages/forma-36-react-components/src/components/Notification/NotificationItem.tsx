import React, { Component } from 'react';
import classNames from 'classnames';
import IconButton from '../IconButton';
import Icon from '../Icon';
import TextLink, { TextLinkProps } from '../TextLink';

import styles from './NotificationItem.css';

export type NotificationIntent = 'success' | 'error' | 'warning';

export interface NotificationCtaProps {
  label: string;
  textLinkProps: Partial<TextLinkProps>;
}

export interface NotificationItemProps {
  intent: NotificationIntent;
  hasCloseButton?: boolean;
  onClose?: Function;
  testId?: string;
  children: React.ReactNode;
  title?: string;
  cta?: Partial<NotificationCtaProps>;
}

const defaultProps = {
  testId: 'cf-ui-notification',
  intent: 'success' as NotificationIntent,
  hasCloseButton: true,
};

export class NotificationItem extends Component<
  NotificationItemProps & typeof defaultProps
> {
  static defaultProps = defaultProps;

  renderTitle() {
    const { title, children } = this.props;

    return (
      <div className={styles.NotificationItem__title}>
        {title ? title : children}
      </div>
    );
  }

  renderBody() {
    const { title, children } = this.props;

    return <div>{title && children}</div>;
  }

  renderCta() {
    const { cta } = this.props;

    if (cta && cta.label)
      return (
        <div>
          <TextLink {...cta.textLinkProps} linkType="white">
            {cta.label}
          </TextLink>
        </div>
      );
  }

  render() {
    const { testId, intent, onClose, hasCloseButton } = this.props;

    const classes = classNames(styles.NotificationItem, {
      [styles[`NotificationItem--${intent}`]]: true,
    });
    return (
      <div
        className={classes}
        data-test-id={testId}
        data-intent={intent}
        role="alert"
        aria-live={intent === 'success' ? 'polite' : 'assertive'}
      >
        <div className={styles.NotificationItem__intent}>{intent}</div>
        <div className={styles.NotificationItem__icon} aria-hidden="true">
          <Icon
            icon={intent === 'success' ? 'CheckCircle' : 'Warning'}
            color="white"
          />
        </div>
        <div className={styles.NotificationItem__text}>
          {this.renderTitle()}
          {this.renderBody()}
          {this.renderCta()}
        </div>
        {hasCloseButton && (
          <IconButton
            buttonType="white"
            iconProps={{ icon: 'Close' }}
            onClick={() => {
              if (onClose) {
                onClose();
              }
            }}
            testId="cf-ui-notification-close"
            label="Dismiss"
            className={styles.NotificationItem__dismiss}
          />
        )}
      </div>
    );
  }
}

export default NotificationItem;
