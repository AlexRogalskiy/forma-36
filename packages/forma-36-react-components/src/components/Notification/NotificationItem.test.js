import React from 'react';
import { shallow, mount } from 'enzyme';
import { axe } from 'jest-axe';
import NotificationItem from './NotificationItem';

it('renders the component', () => {
  const output = shallow(
    <NotificationItem onClose={() => {}} intent="success">
      Notification text
    </NotificationItem>,
  );

  expect(output).toMatchSnapshot();
});

it(`has no a11y issues`, async () => {
  const output = mount(
    <NotificationItem onClose={() => {}} intent="success">
      Notification text
    </NotificationItem>,
  ).html();

  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
