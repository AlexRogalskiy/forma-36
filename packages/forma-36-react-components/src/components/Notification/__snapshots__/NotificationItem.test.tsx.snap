// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div
  aria-live="polite"
  className="NotificationItem NotificationItem--success"
  data-intent="success"
  data-test-id="cf-ui-notification"
  role="alert"
>
  <div
    className="NotificationItem__intent"
  >
    success
  </div>
  <div
    aria-hidden="true"
    className="NotificationItem__icon"
  >
    <Icon
      color="white"
      icon="CheckCircle"
      size="small"
      testId="cf-ui-icon"
    />
  </div>
  <div
    className="NotificationItem__text"
  >
    Notification text
  </div>
  <IconButton
    buttonType="white"
    className="NotificationItem__dismiss"
    hasDropdownCaret={false}
    iconProps={
      Object {
        "icon": "Close",
      }
    }
    isDisabled={false}
    label="Dismiss"
    onClick={[Function]}
    testId="cf-ui-notification-close"
  />
</div>
`;

exports[`renders the component with "error" intent 1`] = `
<div
  aria-live="assertive"
  className="NotificationItem NotificationItem--error"
  data-intent="error"
  data-test-id="cf-ui-notification"
  role="alert"
>
  <div
    className="NotificationItem__intent"
  >
    error
  </div>
  <div
    aria-hidden="true"
    className="NotificationItem__icon"
  >
    <Icon
      color="white"
      icon="Warning"
      size="small"
      testId="cf-ui-icon"
    />
  </div>
  <div
    className="NotificationItem__text"
  >
    Notification text
  </div>
  <IconButton
    buttonType="white"
    className="NotificationItem__dismiss"
    hasDropdownCaret={false}
    iconProps={
      Object {
        "icon": "Close",
      }
    }
    isDisabled={false}
    label="Dismiss"
    onClick={[Function]}
    testId="cf-ui-notification-close"
  />
</div>
`;

exports[`renders the component with "warning" intent 1`] = `
<div
  aria-live="assertive"
  className="NotificationItem NotificationItem--warning"
  data-intent="warning"
  data-test-id="cf-ui-notification"
  role="alert"
>
  <div
    className="NotificationItem__intent"
  >
    warning
  </div>
  <div
    aria-hidden="true"
    className="NotificationItem__icon"
  >
    <Icon
      color="white"
      icon="Warning"
      size="small"
      testId="cf-ui-icon"
    />
  </div>
  <div
    className="NotificationItem__text"
  >
    Notification text
  </div>
  <IconButton
    buttonType="white"
    className="NotificationItem__dismiss"
    hasDropdownCaret={false}
    iconProps={
      Object {
        "icon": "Close",
      }
    }
    isDisabled={false}
    label="Dismiss"
    onClick={[Function]}
    testId="cf-ui-notification-close"
  />
</div>
`;
